Make normals consistent

lindsayad opened the pull request:
Also (perhaps temporarily) adds the option to interpolate the nodal normals or use them straight (e.g. n vs n_j from the Popp literature).

Summary of some convergence results here: https://github.com/idaholab/moose/issues/17539#issuecomment-816011170

@recuero

lindsayad committed:
Update and add to AutomaticMortarGeneration API

- Add `getNodalNormals`
- Add `getNormals`
- Make all data private
- Consequently create some `const` public accessors

lindsayad committed:
Use nodal-based normals in MortarConstraints

We will no longer use normals based solely on the face. We will use the
same nodal normal field that is used to project secondary nodes and
inverse-project primary nodes, so we are consistent. This also adds the
option of whether to interpolate the nodal normal field, e.g. the
traditional way of indexing with quadrature point indices, or to just
use the nodal normals straight and index with the associated dof number.

Closes #17539

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/721045/) on 65226ed wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/17550/site)

This comment will be updated on new commits.

recuero commented on framework/src/constraints/AutomaticMortarGeneration.C:
```diff
@@ -117,6 +116,55 @@ AutomaticMortarGeneration::buildNodeToElemMaps()
 }
 
+std::vector<Point>
+AutomaticMortarGeneration::getNodalNormals(const Elem & secondary_elem) const
+{
+  std::vector<Point> nodal_normals(secondary_elem.n_nodes());
+  for (const auto n : make_range(secondary_elem.n_nodes()))
+    nodal_normals[n] = secondary_node_to_nodal_normal.at(secondary_elem.node_ptr(n));
```
Here, the normals belong to lower-dimensional elements, ie. `secondary_elem` is always a one-dimensional element?

lindsayad replied:
Correct...this should only ever be called with a lower dimensional element from the secondary face

lindsayad replied:
I should perhaps add an assertion that `secondary_elem.dim() == 1`

dschwen replied:
But there is a way forward to do this for 2d secondary elements, right?

lindsayad committed:
Regold all the things

These diffs are generally on the order of 1e-4, 1e-5 which seems
reasonable to me. Antonio and Dewen have reported good (almost surely
better) convergence/results with the update to continuous normals so I
think regolding here is definitely the right thing to do

recuero commented on framework/include/constraints/AutomaticMortarGeneration.h:
```diff
@@ -152,7 +152,64 @@ class AutomaticMortarGeneration : public ConsoleStreamInterface
+  /**
+   * Compute the normals at given reference points on a secondary element
+   * @param secondary_elem The secondary element used to query for associated nodal normals
+   * @param 1d_xi1_pts The reference points on the secondary element to evaluate the normals at. The
```
`1d` -> `oned`

recuero replied:
`1d` -> `oned`

recuero commented on framework/include/constraints/AutomaticMortarGeneration.h:
```diff
@@ -152,7 +152,64 @@ class AutomaticMortarGeneration : public ConsoleStreamInterface
 
-public:
+  /**
+   * @return The nodal normals associated with the provided \p secondary_elem
```
The comments are pretty clear, do we usually add all arguments with `@param` or is it optional?

recuero commented on framework/src/constraints/MortarConstraintBase.C:
```diff
@@ -68,6 +68,12 @@ MortarConstraintBase::validParams()
+      "interpolate_normals",
+      true,
+      "Whether to interpolate the nodal normals (e.g. classic idea of evaluating field at "
+      "quadrature points). If this is set to false, then non-interpolated nodal normals will be "
```
The non-interpolated normal normals are still averaged compared to the element face normals

recuero reviewed (approved):
Looks great to me. Helps with convergence and results consistency 👍

dschwen commented on framework/include/constraints/AutomaticMortarGeneration.h:
```diff
@@ -152,7 +152,64 @@ class AutomaticMortarGeneration : public ConsoleStreamInterface
+   */
+  const std::unordered_multimap<dof_id_type, dof_id_type> & mortarInterfaceCoupling() const
+  {
+    return mortar_interface_coupling;
```
Member variables without a leading underscore. Not introduced by this PR, but maybe you'd like to jump at the opportunity to add another commit to this? :-D

dschwen commented on framework/src/constraints/AutomaticMortarGeneration.C:
```diff
@@ -117,6 +116,55 @@ AutomaticMortarGeneration::buildNodeToElemMaps()
+                                      const std::vector<Real> & oned_xi1_pts) const
+{
+  std::vector<Point> xi1_pts(oned_xi1_pts.size());
+  for (const auto qp : index_range(oned_xi1_pts))
```
`index_range` looks much nicer than that macro wrapped decltype abomination!
